### PR TITLE
Dra 728 implement holdback filter

### PR DESCRIPTION
**Pull request by VictorHarbo**
First implementation of DR holdback filter. Excel sheets are present in aegis for testing. We should discuss where these should be placed on the devel server. (Eventhough the PR says 21 files changed, the 9 of these are test classes)

Hard to test on devel, as the data isnt in preservica stage yet. Please give code a thorough review and comment anything that isn't clear in the documentation as well

BEFORE DOING ANYTHING: Update aegis by

1. Pulling aegis
2. Removing internal_test_files dir in ds-present
3. run kb init in ds-present.

### CODE REVIEW

**Review by thomasegense — CHANGES_REQUESTED** [review]
Looks very solid code and united+embedded solr tested also.
Only major thing is separating all the POI work every time processing a record into a simple java data class and use this when processing the records.

I will delay running with the unit tests for the  review until this refactoring has been submitted.
Also the performance test will be run (another Jira) and this has been solved.

File: src/main/java/dk/kb/present/HoldbackDatePicker.java [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+        return datePicker;
+    }
+
+    /**
+     * Apply the HoldbackDatePicker to a string of XML representing a Preservica Information Object.
+     *
+     * @param record DsStorage record containing an InformationObject from Preservica encapsulating a DR record/program,
+     *              which holdback needs to be calculated for.
+     * @return a string containing the date for when the holdback for the record has expired.
+     *         In the format: yyyy-MM-dd'T'HH:mm:ssZ
```
[the remark this answers is not in the document]
Comment (VictorHarbo) [reply src/main/java/dk/kb/present/HoldbackDatePicker.java:94]: Typo. Normal UTC is also used in the code

File: src/main/java/dk/kb/present/HoldbackDatePicker.java [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+            result.setHoldbackDate("");
+            return result;
+        }
+
+        if (record.getOrigin().equals("ds.tv")){
+            return getHoldbackForTvRecord(record, result);
+        } else if (record.getOrigin().equals("ds.radio")) {
+            return getHoldbackForRadioRecord(record, result);
+        } else {
+            log.error("Holdback cannot be calculated for records that are not from origins 'ds.radio' or 'ds.tv'." +
```
[the remark this answers is not in the document]
Comment (VictorHarbo) [reply src/main/java/dk/kb/present/HoldbackDatePicker.java:110]: Changed to InternalServiceException

File: conf/ds-present-environment.yaml.SAMPLE [hunk 3]
```diff
@@ -53,6 +53,14 @@ licensemodule:
   url: http://localhost:9076/ds-license/v1
   allowall: false
 
+# Absolute paths to Excel sheets containing information used to calculate holdback periods from DR.
+# This value needs to be changed to the correct placement of the files. The placeholder value is resolved from classpath
+# when the files have been injected from aegis. However, full paths can be specief as well.
```
[the remark this answers is not in the document]
Comment (VictorHarbo) [reply conf/ds-present-environment.yaml.SAMPLE:58]: fixed

File: src/main/java/dk/kb/present/HoldbackDatePicker.java [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+            result.setHoldbackDate(addHoldbackDaysToRecordStartDate(xmlStream, 1096));
+            result.setHoldbackPurposeName("");
+            return result;
+        } catch (ParserConfigurationException | SAXException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Calculate holdback for a TV record by comparing values in the record to the DR providede schemas.
```
[the remark this answers is not in the document]
Comment (VictorHarbo) [reply src/main/java/dk/kb/present/HoldbackDatePicker.java:142]: fixed

File: src/main/resources/xslt/preservica2schemaorg.xsl [hunk 5]
```diff
@@ -813,12 +815,30 @@
     <xsl:for-each select="/XIP/Metadata/Content/program_structure:program_structure">
       <xsl:call-template name="program-structure"/>
     </xsl:for-each>
+
+    <!-- Holdback date included here. Holdbac purpose is only included for video objects, therefor it is done in the
```
[the remark this answers is not in the document]
Comment (VictorHarbo) [reply src/main/resources/xslt/preservica2schemaorg.xsl:819]: fixed

File: src/main/java/dk/kb/present/holdback/HoldbackDatePicker.java [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+
+import dk.kb.present.config.ServiceConfig;
+import dk.kb.present.util.saxhandlers.ElementExtractionHandler;
+import dk.kb.storage.model.v1.DsRecordDto;
+import dk.kb.util.DatetimeParser;
+import dk.kb.util.MalformedIOException;
+import dk.kb.util.Resolver;
+import dk.kb.util.webservice.exception.InternalServiceException;
+import org.apache.commons.io.IOUtils;
+import org.apache.poi.ss.usermodel.Cell;
```
Comment (thomasegense) [inline src/main/java/dk/kb/present/holdback/HoldbackDatePicker.java:18]: Clean up unused imports. (POI logic seperated from business logic now)

File: src/main/java/dk/kb/present/holdback/HoldbackDatePicker.java [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+
+    /**
+     * Object representing values from an Excel sheet containing table to find FormNr in based on Form from the XML.
+     */
+    private static FormIndexSheet formIndexSheet;
+
+    /**
+     * Excel sheet containing table to find PurposeID in based on Content and FormNr.
+     */
+    private static XSSFSheet purposeMatrixSheetOld;
```
Comment (thomasegense) [inline src/main/java/dk/kb/present/holdback/HoldbackDatePicker.java:62]: clean up

File: src/main/java/dk/kb/present/holdback/FormIndexSheet.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+import org.apache.poi.ss.usermodel.CellType;
+import org.apache.poi.ss.usermodel.Row;
+import org.apache.poi.xssf.usermodel.XSSFSheet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class FormIndexSheet {
```
Comment (thomasegense) [inline src/main/java/dk/kb/present/holdback/FormIndexSheet.java:14]: Would like a little java class doc.

File: src/main/java/dk/kb/present/holdback/PurposeMatrixSheet.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+import org.apache.poi.ss.usermodel.Row;
+import org.apache.poi.xssf.usermodel.XSSFSheet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+public class PurposeMatrixSheet {
```
Comment (thomasegense) [inline src/main/java/dk/kb/present/holdback/PurposeMatrixSheet.java:15]: Little javadoc wanted

File: src/main/java/dk/kb/present/holdback/PurposeMatrixSheet.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    private static final Logger log = LoggerFactory.getLogger(PurposeMatrixSheet.class);
+
+    private List<Integer> indholdFra = new ArrayList<>();
+    private List<Integer> indholdTil = new ArrayList<>();
+    private List<String> formNr = new ArrayList<>();
+    private final int NUMBER_OF_COLUMNS = 18; // Number of columns in the sheet
```
Comment (thomasegense) [inline src/main/java/dk/kb/present/holdback/PurposeMatrixSheet.java:21]: Can you validate this and throw exception if not correct. Means Excel sheet is out of sync with code.

File: src/main/java/dk/kb/present/holdback/PurposeMatrixSheet.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+            throw new InternalServiceException("There is a difference between the sizes of the lists 'indholdTil' and 'indholdFra'. Their respective sizes are: " + indholdTil.size() + " and " + indholdFra.size());
+        }
+
+        for (List<String> list : formNrValues) {
+            if (list.size() != indholdFra.size()){
+                throw new InternalServiceException("There is a difference between the sizes of a list in the nested list 'formNrValues' and 'indholdFra'. Their respective sizes are: " + list.size() + " and " + indholdFra.size());
+            }
+        }
+
+        log.info("Initialized PurposeMatrixSheet with the following values: \n" +
```
Comment (thomasegense) [inline src/main/java/dk/kb/present/holdback/PurposeMatrixSheet.java:87]: This should be moved to the toString() method  for the class. This is not just a log statement, but a representation of the class.

 And here just call toString() to log it.
Maybe same situation for the other sheet classes.

File: src/main/java/dk/kb/present/webservice/ContextListener.java [hunk 12]
```diff
@@ -67,6 +68,8 @@ public void contextInitialized(ServletContextEvent sce) {
             String configFile = (String) ctx.lookup("java:/comp/env/application-config");
             //TODO this should not refer to something in template. Should we perhaps use reflection here?
             ServiceConfig.initialize(configFile);
+            // Early initialization of HoldbackDatePicker
+            HoldbackDatePicker.init();
```
Comment (thomasegense) [inline src/main/java/dk/kb/present/webservice/ContextListener.java:72]: A few lines down. end bracked } for method not formatted correct. Move to next line. I know it is not part of this review...

**Review by thomasegense — CHANGES_REQUESTED** [review]
Few minor things.

**Pull request merged**